### PR TITLE
[WEB-10065] Refactor client app data loader in PHP lib bootstrap 

### PR DESCRIPTION
**Pull request by AnnieCSerato**
* Remove redundant CLI app, will have to create tech improvement to remove its use from other services(low risk as most don't even know it exists).
* Remove loading local files as thats no longer relevant to the new parsing system.
* Parse new client-applications.json into old array format.
    * If you want to check the old vs new you can run the code on master and add this to the bottom of `parseClientAppData function`
    * `file_put_contents('oldDataArray.txt', print_r($data, true));`
    * and set up a test case to run the application
 ```php
    <?php

namespace Serato\SwsApp\Test\ClientApplication;

use Aws\Sdk;
use Serato\SwsApp\ClientApplication\DataLoader;
use Serato\SwsApp\Test\TestCase;

class Tester extends TestCase
{
    public function testLoadSecrets()
    {
        $dataLoader = new DataLoader(
            'dev',
            new Sdk([
                'region' => 'us-east-1',
                'version' => '2014-11-01',
                'credentials' => [
                    'key' => '<aws key>',
                    'secret' => '<aws secret>'
                ],
            ]),
            $this->getFileSystemCachePool()
        );
        $dataLoader->getApp();
    }
}
```
* Update tests, and remove any redundant tests.
* Minor clean up to remove unused imports and variables

### CODE REVIEW

File: src/ClientApplication/DataLoader.php [hunk 1]
```diff
@@ -214,59 +139,181 @@ private function loadFromS3(string $s3ObjectName): array
...
-                        'Invalid configuration for application `' . $appName . '` in credentials file `' .
-                        $credentialsObjectPath . '`. Missing required key `password_hash`.'
-                    );
-                }
+        foreach ($clientAppsData as $appData) {
+            $parsedData = [];
+            $credentialsData = $this->getSecret($appData['path']);
+            $appSecretName = $this->getSecretName($appData['path'], $this->env);
+            // All apps MUST have `appId`, `appSecret` and `kmsKeyId` keys defined
+            if (!isset($credentialsData['appId'])) {
```
Comment (aosmakserato) [inline src/ClientApplication/DataLoader.php:178]: It is safer to use `empty` here. Isset will return true if the value is an empty string, for example. Same applies to the checks below.

Comment (aosmakserato) [reply src/ClientApplication/DataLoader.php:178]: The cyclomatic complexity of this method is quite high + it feels these methods are a bit out of place. Let's move them into a separate method.

File: src/ClientApplication/DataLoader.php [hunk 2]
```diff
@@ -214,59 +139,181 @@ private function loadFromS3(string $s3ObjectName): array
...
-                // Add optional `kms_key_id` item
-                if (isset($credentialsData[$appName]['kms_key_id'])) {
-                    if (!isset($data[$appName]['jwt'])) {
-                        $data[$appName]['jwt'] = [];
-                    }
-                    $data[$appName]['jwt']['kms_key_id'] = $credentialsData[$appName]['kms_key_id'];
-                }
+            // Add all data, excluding certain keys
+            $parsedData = $appData;
+            unset($parsedData['path'], $parsedData['basic_auth_scopes'], $parsedData['restricted_to']);
```
Comment (aosmakserato) [inline src/ClientApplication/DataLoader.php:199]: Let's add an explanation why we exclude those keys (e.g. "`basic_auth_scopes` has to be renamed to `scopes` in the output")

File: src/ClientApplication/DataLoader.php [hunk 3]
```diff
@@ -214,59 +139,181 @@ private function loadFromS3(string $s3ObjectName): array
...
+                $parsedData['jwt']['access']['restricted_to'] = $appData['restricted_to'];
             }
+
+            array_push($data, $parsedData);
         }
+
         return $data;
     }
+
+    private function getSecretName(string $secretPath, string $env): string
```
Comment (aosmakserato) [inline src/ClientApplication/DataLoader.php:235]: do we even need this method?

**Review by aosmakserato — COMMENTED** [review]
Nice work @AnnieCSerato. I hope you enjoyed this data mapping exercise ;)
I am finished reviewing the dataloader itself, will review the tests a bit later.

File: src/ClientApplication/DataLoader.php [hunk 4]
```diff
@@ -214,59 +139,181 @@ private function loadFromS3(string $s3ObjectName): array
...
+     * @param array $clientAppsData data from client-applications.json
      * @return array
      *
      * @throws MissingApplicationIdException
-     * @throws MissingApplicationPasswordHash
+     * @throws MissingApplicationPassword
+     * @throws MissingKmsKeyIdException
      */
-    private function mergeCredentials(array $commonData, array $credentialsData, string $credentialsObjectPath): array
+    private function parseClientAppData(array $clientAppsData): array
```
Comment (Jasmin12) [inline src/ClientApplication/DataLoader.php:170]: Not sure if we could move the logic of building the parsed data to a separate method to make this methods easier to follow, like 
```
private function parseClientAppData(array $clientAppsData): array
{
    $data = [];
    foreach ($clientAppsData as $appData) {
        $credentialsData = $this->getSecret($appData['path']);
        $parsedData = $this->buildParsedData($appData, $credentialsData);
        array_push($data, $parsedData);
    }
    return $data;
}

```

File: src/ClientApplication/DataLoader.php (hunk 1 above)
Comment (Jasmin12) [inline src/ClientApplication/DataLoader.php:178]: This three validation checks can move to a separate method.

File: src/ClientApplication/DataLoader.php [hunk 5]
```diff
@@ -99,64 +77,11 @@ public function getApp(string $env = null, bool $useCache = true): array
...
-     * @return string
-     */
-    public function getCredentialsObjectName(string $env): string
-    {
-        return str_replace(
-            self::CREDENTIALS_ENV_PLACEHOLDER,
-            $env,
-            self::ENV_CREDENTIALS_NAME_PATTERN
+        return $this->parseClientAppData(
+            $this->loadFromCache(self::COMMON_APP_DATA_NAME, $useCache)
```
Comment (Jasmin12) [inline src/ClientApplication/DataLoader.php:81]: Does the cache last for 1 hour?

File: tests/ClientApplication/DataLoaderTest.php [hunk 6]
```diff
@@ -2,68 +2,76 @@
...
 use Serato\SwsApp\ClientApplication\DataLoader;
 use Serato\SwsApp\Test\TestCase;
-use Exception;
+use Serato\SwsApp\ClientApplication\Exception\InvalidEnvironmentNameException;
+use Serato\SwsApp\ClientApplication\Exception\InvalidFileContentsException;
+use Serato\SwsApp\ClientApplication\Exception\MissingApplicationIdException;
+use Serato\SwsApp\ClientApplication\Exception\MissingApplicationPassword;
+use Serato\SwsApp\ClientApplication\Exception\MissingKmsKeyIdException;
 
 class DataLoaderTest extends TestCase
```
Comment (Jasmin12) [inline tests/ClientApplication/DataLoaderTest.php:13]: Do we need to include the tests to make sure the caching works, like the cache is hit or not?

**Review by Jasmin12 — COMMENTED** [review]
Looks really good @AnnieCSerato ! Amazing work! 👍 👍

**Review by aosmakserato — COMMENTED** [review]
The tests LGTM

File: src/ClientApplication/DataLoader.php (hunk 5 above)
Comment (AnnieCSerato) [reply src/ClientApplication/DataLoader.php:81]: based off the origianl code yes

File: src/ClientApplication/DataLoader.php (hunk 4 above)
Comment (AnnieCSerato) [reply src/ClientApplication/DataLoader.php:170]: This seems but redundant, cause the point of this function is to build the output = the parsed data, maybe have a look again after I've moved the validation of the secrets out of this method which makes the functionality clearer

**Review by Jasmin12 — APPROVED** [review]
Nice one @AnnieCSerato !

**Review by aosmakserato — APPROVED** [review]
Nice work @AnnieCSerato. As I mentioned earlier, we will have to refactor this functionality one more time to fix the performance issues. You have conducted this refactoring according to the VCs and therefore I will be moving the ticket to "Done" and the next iteration of refactoring will be conducted within a separate ticket.

**Pull request merged**